File: frontend/src/models/Invoice.js
import BaseModel from './BaseModel';

export default class Invoice extends BaseModel {
  constructor(data = {}) {
    super(data);
  }

  /**
   * Get invoice ID
   */
  get id() {
    return this._data.invoice_id;
  }

  /**
   * Get creation date
   */
  get createAt() {
    return this._data.create_at;
  }

  /**
   * Get total price
   */
  get totalPrice() {
    return this._data.total_price;
  }

  /**
   * Set total price
   */
  set totalPrice(value) {
    this._data.total_price = value;
  }

  /**
   * Get payment method
   */
  get paymentMethod() {
    return this._data.payment_method;
  }

  /**
   * Set payment method
   */
  set paymentMethod(value) {
    this._data.payment_method = value;
  }

  /**
   * Get staff ID
   */
  get staffId() {
    return this._data.staff_id;
  }

  /**
   * Check if invoice is paid
   */
  get isPaid() {
    return this._data.is_paid;
  }

  /**
   * Set paid status
   */
  set isPaid(value) {
    this._data.is_paid = value;
  }

  /**
   * Create an invoice object from API response
   * @param {Object} data - API data
   */
  static fromApiResponse(data) {
    return new Invoice({
      invoice_id: data.invoice_id,
      create_at: data.create_at,
      total_price: data.total_price,
      payment_method: data.payment_method,
      staff_id: data.staff_id,
      is_paid: data.is_paid
    });
  }
}

File: backend/resource_service/schemas/invoice.py
from datetime import datetime
from pydantic import BaseModel, Field
from typing import Optional

class InvoiceCreate(BaseModel):
    order_id: int = Field(..., description="ID of the order associated with the invoice")
    is_paid: bool = Field(False, description="Indicates if the invoice has been paid")
    total_price: int = Field(..., description="Total price of the invoice")
    
    class Config:
        from_attributes = True
        json_schema_extra = {
            "example": {
                "order_id": 1,
                "total_price": 1000
            }
        }

class InvoiceUpdate(BaseModel):
    is_paid: bool = Field(default=True, description="Indicates if the invoice has been paid")
    create_at: datetime = Field(default= datetime.now(), description="Timestamp when the invoice was paied")
    payment_method: str = Field(..., description="Payment method used for the invoice")
    staff_id: int = Field(..., description="ID of the staff member who created the invoice")

    class Config:
        from_attributes = True
        json_schema_extra = {
            "example": {
                "payment_method": "credit_card",
                "staff_id": 1
            }
        }

class InvoiceResponse(BaseModel):
    invoice_id: int = Field(..., description="ID of the invoice")
    order_id: int = Field(..., description="ID of the order associated with the invoice")
    staff_id: Optional[int] = Field(default=None, description="ID of the staff member who created the invoice")
    create_at: Optional[datetime] = Field(default=None, description="Timestamp when the invoice was paied")
    total_price: int = Field(..., description="Total price of the invoice")
    payment_method: Optional[str] = Field(default=None, description="Payment method used for the invoice")
    is_paid: bool = Field(..., description="Indicates if the invoice has been paid")

    class Config:
        from_attributes = True
        json_schema_extra = {
            "example": {
                "id": 1,
                "order_id": 1,
                "staff_id": 1,
                "create_at": "2023-10-01T12:00:00Z",
                "total_price": 1000,
                "payment_method": "credit_card",
                "is_paid": True
            }
        }